proposals/lip-0054.md

```diff
@@ -301,104 +287,270 @@ messageRecoveryParams = {
+        # The receiving chain must be the terminated chain
+        if ccm.receivingChainID != trs.params.chainID:
+            raise Exception("Cross-chain message receiving chain ID is not valid.")
+        # The sending chain must be live
+        if not isLive(ccm.sendingChainID):
+            raise Exception("Cross-chain message sending chain is not live.")
+
+    # Check the inclusion proof against the sidechain outbox root
+    proof = { 
+        size: terminatedOutboxAccount(trs.params.chainID).outboxSize,
+        idxs: trs.params.idxs,
+        siblingHashes: trs.params.siblingHashes
+    }
+
+    if RMTVerify(
+        [sha256(ccmBytes) for ccmBytes in trs.params.crossChainMessages],
+        proof,
+        terminatedOutboxAccount(trs.params.chainID).outboxRoot
+        ) == False:
+        raise Exception("Message recovery proof of inclusion is not valid.")
 ```
 
-#### Message Recovery Command Execution
+#### Execution
+
+```python
+def execute(trs: Transaction) -> None:
+    senderAddress = sha256(trs.senderPublicKey)[:ADDRESS_LENGTH]
+    # Set CCM status to recovered and assign fee to trs sender
+    crossChainMessages = [decode(crossChainMessageSchema, ccmBytes) for ccmBytes in trs.params.crossChainMessages]
+    recoveredCCMs = []
+
+    for ccmBytes in crossChainMessages:
+        ccmID = sha256(ccmBytes)
+        # Set ccmID as default topic to all events emitted in applyRecovery and forwardRecovery
+        # (instead of the transaction ID).
+        defaultEventsTopic = ccmID
+
+        ccm = decode(crossChainMessageSchema, ccmBytes)
+        # If the sending chain is the mainchain, recover the CCM.
+        # This function never raises an error.
+        if ccm.sendingChainID == CHAIN_ID_MAINCHAIN:
+            applyRecovery(trs, ccm)
+        # If the sending chain is not the mainchain, forward the CCM.
+        # This function never raises an error.
+        elif ccm.sendingChainID != CHAIN_ID_MAINCHAIN:
+            forwardRecovery(trs, ccm)
+        
+        # Append the recovered CCM to the list of recovered CCMs.
+        # Notice that the ccm has been mutated in the applyRecovery and forwardRecovery functions
+        # as the status is set to CCM_STATUS_CODE_RECOVERED (so that it cannot be recovered again).
+        recoveredCCMs.append(encode(crossChainMessageSchema, ccm))
+
+    # Update sidechain outbox root
+    proof = { 
+        "size": terminatedOutboxAccount(trs.params.chainID).outboxSize,
+        "idxs": trs.params.idxs,
+        "siblingHashes": trs.params.siblingHashes
+    }
+
+    terminatedOutboxAccount(trs.params.chainID).outboxRoot = RMTCalculateRoot([sha256(ccmBytes) for ccmBytes in recoveredCCMs], proof)
+```
 
-Processing a transaction `trs` with module ID `MODULE_ID_INTEROPERABILITY` and command ID `COMMAND_ID_MESSAGE_RECOVERY` implies the following logic:
+```python
+def applyRecovery(trs: Transaction, ccm: CCM) -> None:
```

Comment:
I have looked through this function, it looks good. However I haven't checked the details too carefully since I am only roughly aware of the ccm apply logic. I assume, here we just repeat this logic.
The same concerns `forwardRecovery`.

Reply:
the logic is very similar, but we do not terminate the chain in case of errors